### PR TITLE
Fix keyboard calculation for API v21+ by introducing start/stop into EmojiPopup

### PR DESCRIPTION
**Pull request by mario**
Also fixes the issue with the dialog thing.

### CODE REVIEW

**Comment by mario** [comment]
Makes it way way smoother.

**Comment by mario** [comment]
Please test this as much as you can :) Thanks!

**Comment by mario** [comment]
Lint complains of simple things:

[ant:checkstyle] [ERROR] /home/travis/build/[secure]/Emoji/app/src/main/java/com/[secure]/emoji/sample/MainDialog.java:50: Line has trailing spaces. [RegexpSingleline]
1219
1220

Please fix this if that's all that's blocking this PR from being merged.

File: emoji/src/main/java/com/vanniktech/emoji/EmojiPopup.java [hunk 1]
```diff
@@ -214,7 +214,7 @@ void updateKeyboardStateOpened(final int keyboardHeight) {
     final int properWidth = Utils.getProperWidth(context);
 
     if (popupWindow.getWidth() != properWidth) {
-      popupWindow.setWidth(properWidth);E
```
Comment (mario) [inline emoji/src/main/java/com/vanniktech/emoji/EmojiPopup.java]: What was this o.O

**Comment by vanniktech** [comment]
@mario I have no idea how that `E` ended up being there.

**Comment by rubengees** [comment]
I have managed to produce weird behaviour:

- Open Dialog.
- Click Emoji button.
- Close dialog.
- Click Emoji button (normal one).
  - _Emoji popup does not appear._
- Click again.
  - _Weird behaviour (see screenshot)._

device-2019-10-02-185403

Tested on a Pixel 2 API 29 emulator.

**Comment by mario** [comment]
OK @rubengees I know what the issue is. Now on to finding a way to fix it.

**Comment by vanniktech** [comment]
:rocket:

**Comment by mario** [comment]
Please check now. I've identified an issue with emojis only filter (not sure if it was there before) so I'll look into it in a follow-up PR.

File: app/src/main/java/com/vanniktech/emoji/sample/MainActivityMultiAutoCompeteTextView.java [hunk 2]
```diff
@@ -49,7 +49,10 @@
     emojiButton.setColorFilter(ContextCompat.getColor(this, R.color.emoji_icons), PorterDuff.Mode.SRC_IN);
     sendButton.setColorFilter(ContextCompat.getColor(this, R.color.emoji_icons), PorterDuff.Mode.SRC_IN);
 
-    emojiButton.setOnClickListener(ignore -> emojiPopup.toggle());
+    emojiButton.setOnClickListener(ignore -> {
+      emojiPopup.start();
+      emojiPopup.toggle();
```
Comment (vanniktech) [inline app/src/main/java/com/vanniktech/emoji/sample/MainActivityMultiAutoCompeteTextView.java]: Let's call `start()` inside `toggle()`

**Comment by mario** [comment]
Done.

**Comment by vanniktech** [comment]
1. Click on the main edittext in the sample
2. Switch to the emoji keyboard by pressing the left icon
3. Press the back button 
4. The keyboard does get dismissed but the Emoji Dialog is still shown

**Comment by mario** [comment]
Fixed.

File: README.md [hunk 3]
```diff
@@ -98,6 +98,18 @@ emojiPopup.isShowing(); // Returns true when Popup is showing.
...
+```java
+emojiPopup.start()
+```
+
+The following code needs to be executed in onStop():
+
+```java
+emojiPopup.stop()
+```
+
```
Comment (rubengees) [inline README.md]: Can we reuse the `OnAttachStateChangeListener` or add [`ActivityLifecycleCallbacks`](https://developer.android.com/reference/android/app/Application.ActivityLifecycleCallbacks) (we already have an `Activity` context in the popup) to do this on our own to make this easier to use?

Comment (mario) [reply README.md]: We might be able to :P Can you do it? :D

**Comment by mario** [comment]
Thanks for testing @rubengees !

**Comment by mario** [comment]
Btw. in general there's a lots of potential for re-use when using multiple instances of Emoji magic in the same activity - but it's fairly complicated as I need to look where things are referenced and replace them at the proper time.

**Comment by vanniktech** [comment]
I'd also suggest to leave things as they are. (With start/stop needing to be called explicitly)

**Comment by mario** [comment]
Well personally I'd rather remove them, but let's just say I'd rather leave this for later than hassle with it now. 

Btw @vanniktech sent you an email yesterday.

**Comment by vanniktech** [comment]
Agreed and I already replied ;)

**Comment by rubengees** [comment]
I would propose to do this before the next release though, because that makes the library easier to use (and we will get less bug reports of people forgetting to call these methods 😀).

File: README.md (hunk 3 above)
Comment (rubengees) [reply README.md]: I'll try 🙂

**Comment by mario** [comment]
Thanks folks!

**Pull request merged**